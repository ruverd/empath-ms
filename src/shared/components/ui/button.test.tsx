import { render, screen } from "@testing-library/react";
import { Mail } from "lucide-react";
import { describe, expect, it } from "vitest";

import { Button } from "./button";

describe("Button", () => {
  it("should render with default variant", () => {
    render(<Button>Click me</Button>);

    const button = screen.getByRole("button", { name: "Click me" });
    expect(button).toBeInTheDocument();
  });

  it("should render all variants correctly", () => {
    const variants = ["secondary", "outline", "ghost", "link"] as const;

    variants.forEach((variant) => {
      render(<Button variant={variant}>{variant} button</Button>);
      const button = screen.getByRole("button", { name: `${variant} button` });

      switch (variant) {
        case "secondary":
          expect(button).toHaveClass(
            "bg-secondary",
            "text-secondary-foreground"
          );
          break;
        case "outline":
          expect(button).toHaveClass(
            "border-secondary-aqua",
            "text-secondary-aqua",
            "bg-primary-indigo"
          );
          break;
        case "ghost":
          expect(button).toHaveClass(
            "hover:bg-accent",
            "hover:text-primary-cloud"
          );
          break;
        case "link":
          expect(button).toHaveClass("text-primary");
          break;
      }
    });
  });

  it("should render all sizes correctly", () => {
    const sizes = ["default", "xs", "sm", "lg", "icon"] as const;

    sizes.forEach((size) => {
      render(<Button size={size}>{size} button</Button>);
      const button = screen.getByRole("button", { name: `${size} button` });

      switch (size) {
        case "default":
          expect(button).toHaveClass("h-10", "px-4", "py-2");
          break;
        case "xs":
          expect(button).toHaveClass("h-8", "px-4");
          break;
        case "sm":
          expect(button).toHaveClass("h-9", "px-4");
          break;
        case "lg":
          expect(button).toHaveClass("h-11", "px-8");
          break;
        case "icon":
          expect(button).toHaveClass("h-10", "w-10");
          break;
      }
    });
  });

  it("should render in active state", () => {
    render(<Button isActive>Active Button</Button>);

    const button = screen.getByRole("button", { name: "Active Button" });
    expect(button).toHaveClass("bg-accent", "text-accent-foreground");
  });

  it("should render as disabled", () => {
    render(<Button disabled>Disabled Button</Button>);

    const button = screen.getByRole("button", { name: "Disabled Button" });
    expect(button).toBeDisabled();
    expect(button).toHaveClass(
      "disabled:pointer-events-none",
      "disabled:opacity-50"
    );
  });

  it("should render with icon", () => {
    render(
      <Button>
        <Mail data-testid="mail-icon" />
        With Icon
      </Button>
    );

    const button = screen.getByRole("button", { name: "With Icon" });
    const icon = screen.getByTestId("mail-icon");

    expect(button).toContainElement(icon);
    expect(button).toHaveClass("[&_svg]:size-4", "[&_svg]:shrink-0");
  });

  it("should apply custom className", () => {
    const customClass = "custom-class";
    render(<Button className={customClass}>Custom Button</Button>);

    const button = screen.getByRole("button", { name: "Custom Button" });
    expect(button).toHaveClass(customClass);
  });

  it("should render as child component when asChild is true", () => {
    render(
      <Button asChild>
        <a href="#">Link Button</a>
      </Button>
    );

    expect(screen.getByRole("link")).toBeInTheDocument();
    expect(screen.queryByRole("button")).not.toBeInTheDocument();
  });

  it("should match snapshot with default variant", () => {
    const { container } = render(<Button>Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should match snapshot with icon and secondary variant", () => {
    const { container } = render(
      <Button variant="secondary">
        <Mail data-testid="mail-icon" />
        With Icon
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should match snapshot when disabled", () => {
    const { container } = render(<Button disabled>Disabled Button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
