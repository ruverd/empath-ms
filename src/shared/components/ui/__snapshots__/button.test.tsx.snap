// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should match snapshot when disabled 1`] = `
<button
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-[40px] text-sm font-bold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
  disabled=""
>
  Disabled Button
</button>
`;

exports[`Button > should match snapshot with default variant 1`] = `
<button
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-[40px] text-sm font-bold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
>
  Click me
</button>
`;

exports[`Button > should match snapshot with icon and secondary variant 1`] = `
<button
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-[40px] text-sm font-bold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 px-4 py-2"
>
  <svg
    class="lucide lucide-mail"
    data-testid="mail-icon"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      height="16"
      rx="2"
      width="20"
      x="2"
      y="4"
    />
    <path
      d="m22 7-8.97 5.7a1.94 1.94 0 0 1-2.06 0L2 7"
    />
  </svg>
  With Icon
</button>
`;
